SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/LegacyLoadBalancingStrategy.cpp

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "LoadBalancing/LegacyLoadBalancingStrategy.h"
+
+#include "EngineClasses/SpatialVirtualWorkerTranslator.h"
+#include "Interop/Connection/SpatialOSWorkerInterface.h"
+#include "LoadBalancing/AbstractLBStrategy.h"
+#include "LoadBalancing/LBDataStorage.h"
+#include "LoadBalancing/LoadBalancingCalculator.h"
+#include "LoadBalancing/PartitionManager.h"
+
+namespace SpatialGDK
+{
+FLegacyLoadBalancing::FLegacyLoadBalancing(UAbstractLBStrategy& LegacyLBStrat, SpatialVirtualWorkerTranslator& InTranslator)
+	: Translator(InTranslator)
+{
+	ExpectedWorkers = LegacyLBStrat.GetMinimumRequiredWorkers();
+	VirtualWorkerIdToHandle.SetNum(ExpectedWorkers);
+
+	if (!LegacyLBStrat.IsStrategyWorkerAware())
+	{
+		AssignmentStorage = MakeUnique<FDirectAssignmentStorage>();
+	}
+	else
+	{
+		PositionStorage = MakeUnique<SpatialGDK::FSpatialPositionStorage>();
+		GroupStorage = MakeUnique<SpatialGDK::FActorGroupStorage>();
+
+		FLegacyLBContext LBContext;
+		Calculator = LegacyLBStrat.CreateLoadBalancingCalculator(LBContext);
+
+		if (LBContext.Layers != nullptr)
+		{
+			LBContext.Layers->SetGroupData(*GroupStorage);
+		}
+		for (auto* Grid : LBContext.Grid)
+		{
+			Grid->SetPositionData(*PositionStorage);
+		}
+	}
+}
+
+FLegacyLoadBalancing::~FLegacyLoadBalancing() {}
+
+void FLegacyLoadBalancing::Advance(ISpatialOSWorker& Connection)
+{
+	const TArray<Worker_Op>& Messages = Connection.GetWorkerMessages();
+
+	for (const auto& Message : Messages)
+	{
+		switch (Message.op_type)
+		{
+		case WORKER_OP_TYPE_ENTITY_QUERY_RESPONSE:
+		{
+			const Worker_EntityQueryResponseOp& Op = Message.op.entity_query_response;
+			if (WorkerTranslationRequest.IsSet() && Op.request_id == WorkerTranslationRequest.GetValue())
+			{
+				if (Op.status_code == WORKER_STATUS_CODE_SUCCESS)
+				{
+					for (uint32_t i = 0; i < Op.results[0].component_count; i++)
+					{
+						Worker_ComponentData Data = Op.results[0].components[i];
+						if (Data.component_id == SpatialConstants::VIRTUAL_WORKER_TRANSLATION_COMPONENT_ID)
+						{
+							Schema_Object* ComponentObject = Schema_GetComponentDataFields(Data.schema_type);
+							Translator.ApplyVirtualWorkerManagerData(ComponentObject);
+							bTranslatorIsReady = true;
+							for (uint32 VirtualWorker = 0; VirtualWorker < ExpectedWorkers; ++VirtualWorker)
+							{
+								if (Translator.GetServerWorkerEntityForVirtualWorker(VirtualWorker + 1)
+									== SpatialConstants::INVALID_ENTITY_ID)
+								{
+									bTranslatorIsReady = false;
+								}
+							}
+						}
+					}
+				}
+			}
+			WorkerTranslationRequest.Reset();
```

Comment:
This will get reset on any query response, when it should only be in response to the desired query response.